File: app/src/main/java/knf/kuma/backup/objects/SeenList.java
package knf.kuma.backup.objects;

import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;
import com.google.gson.reflect.TypeToken;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import knf.kuma.database.CacheDB;
import knf.kuma.database.dao.AnimeDAO;
import knf.kuma.pojos.AnimeObject;
import xdroid.toaster.Toaster;

/**
 * Created by jordy on 01/03/2018.
 */

public class SeenList {
    @SerializedName("response")
    String response;
    @SerializedName("vistos")
    String vistos;
    List<SeenObj> list;

    public static List<AnimeObject.WebInfo.AnimeChapter> decode(InputStream inputStream) {
        int errorCount = 0;
        AnimeDAO dao = CacheDB.INSTANCE.animeDAO();
        SeenList seenList = new Gson().fromJson(new InputStreamReader(inputStream), new TypeToken<SeenList>() {
        }.getType());
        seenList.deserialize();
        int totalCount = seenList.list.size();
        List<AnimeObject.WebInfo.AnimeChapter> chapters = new ArrayList<>();
        AnimeObject animeObject = null;
        for (SeenObj obj : seenList.list) {
            if (animeObject == null || !animeObject.aid.equals(obj.aid))
                animeObject = dao.getByAid(obj.aid);
            List<AnimeObject.WebInfo.AnimeChapter> chapterList = animeObject.chapters;
            boolean found = false;
            for (AnimeObject.WebInfo.AnimeChapter chapter : chapterList) {
                if (chapter.number.endsWith(" " + obj.num)) {
                    chapters.add(chapter);
                    found = true;
                    break;
                }
            }
            if (!found)
                errorCount++;
        }
        Toaster.toast("Migrados correctamente " + (totalCount - errorCount) + "/" + totalCount);
        return chapters;
    }

    private void deserialize() {
        list = new ArrayList<>();
        Log.e("Seen", vistos);
        String[] els = vistos.replace("E", "").split(":::");
        for (String el : els) {
            if (!el.equals("")) {
                String[] spl = el.split("_");
                list.add(new SeenObj(spl[0], spl[1]));
            }
        }
        Collections.sort(list);
    }

    class SeenObj implements Comparable<SeenObj> {
        String aid;
        String num;

        SeenObj(String aid, String num) {
            this.aid = aid;
            this.num = num;
        }

        @Override
        public int compareTo(@NonNull SeenObj o) {
            int bname = aid.compareTo(o.aid);
            if (bname != 0) {
                return bname;
            } else {
                return num.compareTo(o.num);
            }
        }
    }
}

File: app/src/main/java/knf/kuma/backup/objects/SeenList.kt
package knf.kuma.backup.objects

import android.util.Log
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.google.gson.reflect.TypeToken
import knf.kuma.database.CacheDB
import knf.kuma.pojos.SeenObject
import xdroid.toaster.Toaster
import java.io.InputStream
import java.io.InputStreamReader
import java.util.*

/**
 * Created by jordy on 01/03/2018.
 */

class SeenList {
    @SerializedName("response")
    internal var response: String? = null
    @SerializedName("vistos")
    internal var vistos: String? = null
    internal var list: MutableList<SeenObj>? = null

    private fun deserialize() {
        list = ArrayList()
        Log.e("Seen", vistos)
        val els = vistos?.replace("E", "")?.split(":::".toRegex())?.dropLastWhile { it.isEmpty() }?.toTypedArray()
        for (el in els ?: emptyArray()) {
            if (el != "") {
                val spl = el.split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                list?.add(SeenObj(spl[0], spl[1]))
            }
        }
        list?.sort()
    }

    internal inner class SeenObj(var aid: String, var num: String) : Comparable<SeenObj> {

        override fun compareTo(other: SeenObj): Int {
            val bname = aid.compareTo(other.aid)
            return if (bname != 0) {
                bname
            } else {
                num.compareTo(other.num)
            }
        }
    }

    companion object {

        fun decode(inputStream: InputStream?): MutableList<SeenObject>? {
            if (inputStream == null) return null
            var errorCount = 0
            val dao = CacheDB.INSTANCE.animeDAO()
            val seenList = Gson().fromJson<SeenList>(InputStreamReader(inputStream), object : TypeToken<SeenList>() {

            }.type)
            seenList.deserialize()
            val totalCount = seenList.list?.size ?: 0
            val chapters = ArrayList<SeenObject>()
            var animeObject: AnimeChapters? = null
            for (obj in seenList.list ?: listOf<SeenObj>()) {
                try {
                    if (animeObject == null || animeObject.aid != obj.aid)
                        animeObject = dao.getChaptersByAid(obj.aid)
                    val chapterList = animeObject.chaptersList()
                    var found = false
                    for (chapter in chapterList) {
                        try {
                            if (chapter.number.endsWith(" " + obj.num)) {
                                chapters.add(SeenObject.fromChapter(chapter))
                                found = true
                                break
                            }
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                    if (!found)
                        errorCount++
                } catch (e: Exception) {
                    errorCount++
                }
            }
            Toaster.toast("Migrados correctamente " + (totalCount - errorCount) + "/" + totalCount)
            return chapters
        }
    }
}
